server/util/rpcutil/rpcutil.go

```diff
@@ -84,7 +84,11 @@ type Sender[T proto.Message] struct {
 // acknowledge individual messages. A timeout wil only occur if the sender
 // exhausts the flow-control window and the receiver does not increase it.
 func (s *Sender[T]) SendWithTimeoutCause(msg T, timeout time.Duration, cause error) error {
-	s.sendChan <- msg
+	select {
```

Comment:
I'm having a hard time seeing how this can occur in practice unless there's a bug in the caller. 
`sendChan` is a buffered channel of size 1 which means up to 1 item can be enqueued without blocking. 
If this function returns with an error, SendWithTimeoutCause should not be called again. If it returns without a `nil` error it means sendChan has been drained and will not block.

Reply:
I just spent some time to provoke a hang after canceling the context passed to the `Sender` and could not. The hang I saw yesterday was likely a bug in the Writer I was working on.

> If this function returns with an error, SendWithTimeoutCause should not be called again.

I've explicitly worked that constraint into the Writer I'm working on.